### DISCUSSION TITLE
Localized Grid Reductions

### DISCUSSION THREAD

**Discussion by philipc2** [UXARRAY/uxarray#724] [Ideas]
## Overview

Currently implemented is the `UxDataArray.nodal_average()` function, which takes the average of all the nodes that surrounds each face. 

We are looking to generalize this implementation and support various operations (`mean`, `max`, `min`, `sum`, etc.) and across different grid elements. 

Additionally, we would like to ensure that our **naming of this functionality** doesn't lead to any confusion and is consistent with literature and other applications.  INPUT NEEDED FROM COMMUNITY

## Operations

Below is a list of intended operators to support:
* `mean`
* `max`
* `min`
* `prod`
* `sum`
* `std`
* `var`
* `median`
* `all`
* `any`

These operations (when applied to the whole data) are typically referred to as "reductions", since applying them reduces the dimensionality of the data. However, when we consider the connectivity information of the grid, and apply the reductions on pairings of elements (as described in detail below), the resulting dimension will always either be `n_node`, `n_edge`, or `n_face` 

## Localized Reductions

Continuing with the Nodal Average example, we can consider this to be a localized reduction performed at each **face**.  Applying the `mean` operation on the nodes that surround each face would transform our node centered data variable into a face centered data variable. 

Speaking more generally, by applying these reductions on each element using connectivity information, we can transform any data variable from one data mapping to another. Below is a list of all possible pairings:
* Node to Face
* Node to Edge
* Edge to Face
* Edge to Node
* Face to Node
* Face to Edge

Below are examples of each of these pairings. 

### Node to Face

Operation is applied to all the nodes that surround each face, with the result being stored on the faces. 

`face_node_connectivity` is used to index the node-centered data variable.

![node_face](https://github.com/UXARRAY/uxarray/assets/67855069/6b1d4435-4f9a-40b3-8cb2-22aa2d4785be)

### Node to Edge

Operation is applied to the two nodes that make up each edge, with the result stored on the edges.

`edge_node_connectivity` is used to index the node-centered data variable.

![node_edge](https://github.com/UXARRAY/uxarray/assets/67855069/1d9d2ca8-e245-48c7-ac74-83cb8981f066)

### Edge to Face

Operation is applied to the edges that surround each face, with the result stored on the faces.

`face_edge_connectivity` is used to the index the edge-centered data variable.

![edge_face](https://github.com/UXARRAY/uxarray/assets/67855069/d8800604-b924-43b7-a21c-263faa291e8d)

### Edge to Node

Operation is applied to the edges that surround each node, with the result stored on the nodes.

`node_edge_connectivity` is used to index the edge-centered data variable.

![edge_node](https://github.com/UXARRAY/uxarray/assets/67855069/1a8205c9-d337-4a04-8eeb-63d48e7ea93f)

### Face to Node

Operation is applied to the faces that surround each node, with the result stored on the nodes

`node_face_connectivity` is used to index the face-centered data variable

![face_node](https://github.com/UXARRAY/uxarray/assets/67855069/fc70a65c-5b14-4b2b-ac37-02bf35d67e1f)

### Face to Edge

Operation is applied to the faces that saddle each edge, with the result stored on the edges.

`edge_face_connectivity` is used to index the face-centered data variable.

![face_edge](https://github.com/UXARRAY/uxarray/assets/67855069/07abf62e-3267-4d74-8506-58d6557a5458)

**Comment by philipc2** [comment]
The term that I've used to describe these operations is "Localized" or "Grid-Informed" reductions, since instead of applying the reduction the entire dataset at once, the reductions are performed on each element. 

We've also floated the idea of "Spatial" reductions, but are not entirely sure which one would be the most suitable for this. 

I'd like to see what input anyone has, especially any "proper" way to refer to these operations.

**Reply by philipc2** [reply]
Some proof-of-concept implementations for a node to face reduction of a node-centered data variable.

```Python
# regular xarray mean for reference, operation is applied to ALL values globally
UxDataArray.mean(dim='n_node')

# use a parameter to indicate a local/topological/spatial mean
UxDataArray.mean(dim='n_node', localized=True, destination='faces')
UxDataArray.mean(dim='n_node', spatial=True, destination='faces')
UxDataArray.mean(dim='n_node', grid_aware=True, destination='faces')
UxDataArray.mean(dim='n_node', grid_informed=True, destination='faces')

# separate methods 
UxDataArray.spatial_mean(destination='faces')
UxDataArray.localized_mean(destination='faces')

# methods grouped under an accessor
UxDataArray.spatial.mean(destination='faces')
````

Any preferences? I personally like the `spatial` accessor for grouping these reductions, and it could be used in the future for grouping other operations too.

**Comment by philipc2** [comment]
Pinging @brianpm @rljacob @paullric @erogluorhan

**Comment by iulian787** [comment]
the above definitions are pure "topological" in nature, in my opinion;  they use just the "connectivity" information

for averages, there could be also some "weighted" averages defined for nodes, edges, faces, based on multiple things
for example, for a node average computed from connected faces, the weight could be the corresponding angle or the corresponding face area;  or from an edge to node, the length of the edge could be considered
so another parameter could be added (..., weight='area',... ) or weight='length', 'angle', 

"spatial" average should refer in my opinion to some sort of distance from the node to the center of the faces that fall within a radius; again, some sort of weighting based on the distance should be involved ( RBF is radial basis function , etc). Also, what is the "center" of a face ? Most likely is the "center of mass" for that face, but some might use the center of the circumscribed circle in the Delaunay triangulation? I would not go there, I would use always "center of mass". For an edge, the midpoint of the edge should be its "center" 

what would "sum" mean locally ? I would not support that, except in an integral way, like total mass in a conservation paradigm, globally, not locally 

min, max, also does not make much sense locally

**Reply by philipc2** [reply]
Hi @iulian787 

Thank you so much for the feedback! 

> the above definitions are pure "topological" in nature, in my opinion; they use just the "connectivity" information

Referring to them as topological makes a lot of sense in this context. 

> for averages, there could be also some "weighted" averages defined for nodes, edges, faces, based on multiple things
> for example, for a node average computed from connected faces, the weight could be the corresponding angle or the corresponding face area; or from an edge to node, the length of the edge could be considered
> so another parameter could be added (..., weight='area',... ) or weight='length', 'angle',

This is indeed something we've been looking to introduce.

> "spatial" average should refer in my opinion to some sort of distance from the node to the center of the faces that fall within a radius; again, some sort of weighting based on the distance should be involved ( RBF is radial basis function , etc). 

This makes sense. We have some remapping functionality (nearest neighbor, inverse distance weighted) that uses a KD/Ball tree to obtain these distances, which is in line with this.

> Also, what is the "center" of a face ? Most likely is the "center of mass" for that face, but some might use the center of the circumscribed circle in the Delaunay triangulation? I would not go there, I would use always "center of mass".

When the centers aren't provided in the grid file, we compute them by taking the Cartesian average. 

> For an edge, the midpoint of the edge should be its "center"

This is what we do.

> what would "sum" mean locally ? I would not support that, except in an integral way, like total mass in a conservation paradigm, globally, not locally
> 
> min, max, also does not make much sense locally

One application of min and max locally would be to quicky turn a node-centered or edge-centered data variable into a face-centered one for polygon visualization, which would preserve the local minimums and maximums for each given face, something that would be lost if we did a mean operation for the same purpose
